### PR TITLE
[ISSUE #180]add tag filter example.

### PR DESCRIPTION
**Pull request by xujianhai666**
## What is the purpose of the change

ISSUE #180 add tag filter example, fix heartbeat.

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

### CODE REVIEW

**Comment by ShannonDing** [comment]
Test failed：
# github.com/apache/rocketmq-client-go/consumer [github.com/apache/rocketmq-client-go/consumer.test]
consumer/consumer_test.go:73:29: not enough arguments in call to rmqCli.EXPECT().InvokeSync
	have (gomock.Matcher, gomock.Matcher, gomock.Matcher)
	want (interface {}, interface {}, interface {}, interface {})
consumer/consumer_test.go:130:30: not enough arguments in call to rmqCli.EXPECT().InvokeSync
	have (gomock.Matcher, gomock.Matcher, gomock.Matcher)
	want (interface {}, interface {}, interface {}, interface {})
consumer/consumer_test.go:158:30: not enough arguments in call to rmqCli.EXPECT().InvokeSync
	have (gomock.Matcher, gomock.Matcher, gomock.Matcher)
	want (interface {}, interface {}, interface {}, interface {})
FAIL	github.com/apache/rocketmq-client-go/consumer [build failed]

File: consumer/push_consumer.go [hunk 1]
```diff
@@ -493,6 +493,8 @@ func (pc *pushConsumer) pullMessage(request *PullRequest) {
 			goto NEXT
 		}
 
+		pc.processPullResult(request.mq, result, sd)
```
Comment (wenfengwang) [inline consumer/push_consumer.go]: the L507 will overwrite result of `pc.processPullResult(request.mq, result, sd)`, could you delete L507 and retest？

File: primitive/message.go [hunk 2]
```diff
@@ -85,28 +85,33 @@ func NewMessage(topic string, body []byte) *Message {
 	return msg
 }
 
-// SetDelayTimeLevel set message delay time to consume.
+// WithDelayTimeLevel set message delay time to consume.
 // reference delay level definition: 1s 5s 10s 30s 1m 2m 3m 4m 5m 6m 7m 8m 9m 10m 20m 30m 1h 2h
 // delay level starts from 1. for example, if we set param level=1, then the delay time is 1s.
-func (msg *Message) SetDelayTimeLevel(level int) {
+func (msg *Message) WithDelayTimeLevel(level int) *Message {
```
Comment (ShannonDing) [inline primitive/message.go]: modify the API signature should be more careful. IMO, it is better to add a new API here and leave the old one with Deprecated flag and then remove it.

File: consumer/push_consumer.go (hunk 1 above)
Comment (xujianhai666) [reply consumer/push_consumer.go]: dome

File: primitive/message.go (hunk 2 above)
Comment (xujianhai666) [reply primitive/message.go]: becase of the alpha now, so maybe okay here. when we release, we should be careful.

**Comment by xujianhai666** [comment]
> Test failed：
> #github.com/apache/rocketmq-client-go/consumer [github.com/apache/rocketmq-client-go/consumer.test]
> consumer/consumer_test.go:73:29: not enough arguments in call to rmqCli.EXPECT().InvokeSync
> have (gomock.Matcher, gomock.Matcher, gomock.Matcher)
> want (interface {}, interface {}, interface {}, interface {})
> consumer/consumer_test.go:130:30: not enough arguments in call to rmqCli.EXPECT().InvokeSync
> have (gomock.Matcher, gomock.Matcher, gomock.Matcher)
> want (interface {}, interface {}, interface {}, interface {})
> consumer/consumer_test.go:158:30: not enough arguments in call to rmqCli.EXPECT().InvokeSync
> have (gomock.Matcher, gomock.Matcher, gomock.Matcher)
> want (interface {}, interface {}, interface {}, interface {})
> FAIL	github.com/apache/rocketmq-client-go/consumer [build failed]

done

File: consumer/consumer.go [hunk 3]
```diff
@@ -799,10 +799,10 @@ func (dc *defaultConsumer) processPullResult(mq *primitive.MessageQueue, result
 
 		// filter message according to tags
 		msgListFilterAgain := msgs
-		if len(data.Tags) > 0 && data.ClassFilterMode {
+		if data.Tags.Len() > 0 && data.ClassFilterMode {
```
Comment (ShannonDing) [inline consumer/consumer.go]: Why make msgListFilterAgain a length of len(msgs).
Maybe it should be initialized to zero.

Comment (xujianhai666) [reply consumer/consumer.go]: done @ShannonDing

**Pull request merged**